File: clients/intellij-community/src/main/java/dev/cleanroommc/workbench/intellij/community/CurrentContextFeatureClient.java
package dev.cleanroommc.workbench.intellij.community;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.List;
import java.util.Set;

/** Product-owned path/identity-free action port for the selected Work Session. */
public final class CurrentContextFeatureClient {
    private static final Set<String> ACTIONS = Set.of(
            "open", "test", "apply", "verify", "rollback", "recover"
    );
    private static final int MAXIMUM_OUTPUT_BYTES = 48 * 1024 * 1024;

    private CurrentContextFeatureClient() {
    }

    public static @NotNull List<String> arguments(@NotNull String action) {
        if (!ACTIONS.contains(action)) {
            throw new IllegalArgumentException(
                    "current Work Session feature action is unsupported"
            );
        }
        return List.of("change", "material-fluid-recipe", action, "--json");
    }

    public static @NotNull CommandProcess.Observation invoke(
            @NotNull CoreLaunch launch,
            @NotNull String action,
            long timeoutSeconds,
            @Nullable String ambientWorkingDirectory
    ) throws IOException {
        CommandProcess.Observation observation = CommandProcess.captureObserved(
                launch,
                arguments(action),
                MAXIMUM_OUTPUT_BYTES,
                timeoutSeconds,
                ambientWorkingDirectory
        );
        String stderr = observation.stderr().trim();
        if (observation.exitCode() != 0 || !stderr.isEmpty()) {
            String detail = stderr.isEmpty()
                    ? observation.stdout().trim()
                    : stderr;
            throw new IOException(
                    "current Work Session action exited " + observation.exitCode()
                            + (detail.isEmpty()
                            ? ""
                            : ":\n" + detail.substring(
                                    0, Math.min(16 * 1024, detail.length())
                            ))
            );
        }
        return observation;
    }
}

File: clients/vscode/currentContextFeatureClient.js
"use strict";

const { invokeCoreJson } = require("./coreCommandClient");

const ACTIONS = new Set(["open", "test", "apply", "verify", "rollback", "recover"]);
const FORMAT = "workbench-current-context-feature-action-v1";

function currentContextFeatureArguments(action) {
  if (typeof action !== "string" || !ACTIONS.has(action)) {
    throw new Error("current Work Session feature action is unsupported");
  }
  return Object.freeze([
    "change", "material-fluid-recipe", action, "--json",
  ]);
}

/** Invoke the selected Work Session context without accepting owner IDs or paths. */
async function invokeCurrentContextFeatureAction(executable, action, options = {}) {
  const arguments_ = currentContextFeatureArguments(action);
  let invocation = null;
  const outcome = await invokeCoreJson(executable, arguments_, {
    ...options,
    label: `Workbench current-context ${action}`,
    observeInvocation: (value) => { invocation = value; },
  });
  if (invocation === null) {
    throw new Error("current Work Session feature action lacks process custody");
  }
  return Object.freeze({
    action,
    arguments: arguments_,
    format: FORMAT,
    invocation,
    outcome,
    schema_version: 1,
  });
}

module.exports = {
  ACTIONS,
  FORMAT,
  currentContextFeatureArguments,
  invokeCurrentContextFeatureAction,
};
